torch/fx/traceback.py

```diff
@@ -15,12 +20,76 @@
     "format_stack",
     "set_current_meta",
     "get_current_meta",
+    "NodeSource",
+    "get_graph_provenance_json",
 ]
 
 current_meta: Dict[str, Any] = {}
 should_preserve_node_meta = False
 
 
+class NodeSource:
+    """
+    NodeSource is a data structure that contains the provenance information of a node.
+    If node `a` is created from node `b`, then `a.meta["from_node"]` may contain NodeSource(b).
+    """
+
+    node_name: str
+    target: str
+    graph_id: int
+    pass_name: str
+    action: str
+    from_node: List["NodeSource"]
+
+    def __init__(self, node: Optional[Node], pass_name: str = "", action: str = ""):
+        self.pass_name = pass_name
+        self.action = action
+
+        if action not in ["create", "replace", ""]:
+            warnings.warn(f"unknown action for node source, {action}")
+
+        if node:
+            self.node_name = node.name
+            self.target = str(node.target)
+            self.graph_id = id(node.graph)
+            self.from_node = (
+                copy.deepcopy(node.meta["from_node"])
+                if "from_node" in node.meta
+                else []
+            )
+        else:
+            self.node_name = ""
+            self.target = ""
+            self.graph_id = -1
+            self.from_node = []
+
+    def __repr__(self):
+        return self.print_readable()
+
+    def print_readable(self, indent=0):
+        if indent > 36:
```

Comment:
Feels a bit arbitrary? In any case I'd probably keep `indent` = 0, 1, ... and multiply by 4 when creating the string so that you can guard on the level instead of 4 x level.

Reply:
Fixed now!